### DISCUSSION TITLE
[Web API] 'ResizeObserver' For callback functions related, is it called immediately after 'reflow' or after 'paint'?

### DISCUSSION THREAD

**Discussion by MTtankkeo** [community/community#135212] [Programming Help] [answered]
### Body

I understand that the ResizeObserver is called not only when the size of the element changes but also when the initial size is defined.

So, that made me curious.
Is the callback function related to ResizeObserver called immediately after the 'reflow' or only after the 'paint' step?

In other words, is the element measured and the callback invoked immediately after that, or is it called after one frame?

> The main reason I'm asking is that I need to create a custom element that's closely tied to the rendering of the element. If the size changes, I need to perform certain actions accordingly, but if those actions are triggered after the paint step, I worry that the user might notice some flickering.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by everforyou** [comment, marked as answer]
Hello,

The `ResizeObserver` callback is invoked after layout (reflow) but before the paint phase.

Here’s how it works:
1. **Reflow (Layout):** When an element's size changes, the browser performs a reflow to recalculate the positions and sizes of elements.
2. **ResizeObserver Callback:** Once the layout calculations are done, the `ResizeObserver` triggers its callback with the updated size of the observed elements. This occurs before the paint phase.
3. **Paint:** After the `ResizeObserver` callback runs, the browser proceeds to paint the elements.

Thus, the callback is executed after reflow but before paint, allowing you to handle size changes without causing flickering or other visual issues as the user described.

For your use case, this behavior ensures that any actions you perform in the `ResizeObserver` callback will happen before the browser has rendered the updated UI to the screen, reducing the likelihood of flickering.

**Reply by MTtankkeo** [reply]
Thanks, It helped me a lot! 😀
